File: dpm/installer/lockfile.py
"""
lockfile.py - for reading and writing dpm.lock files
"""

import json
from pathlib import Path
from typing import Dict, List, Optional
from datetime import datetime


class LockFile:
    """for reading and writing dpm.lock files"""
    
    def __init__(self, lockfile_path: str = "dpm.lock"):
        self.lockfile_path = Path(lockfile_path)
    
    def exists(self) -> bool:
        """check if lock file exists"""
        return self.lockfile_path.exists()
    
    def read(self) -> Optional[Dict]:
        """read lock file"""
        if not self.exists():
            return None
        
        try:
            with open(self.lockfile_path, 'r') as f:
                return json.load(f)
        except Exception:
            return None
    
    def write(self, selected_versions: Dict[str, str],
              dependency_map: Dict[str, List[str]],
              package_info: Dict[str, Dict[str, str]]):
        """write lock file with resolved versions"""
        lock_data = {
            "version": "1.0",
            "generated": datetime.now().isoformat(),
            "packages": {}
        }
        
        for name, version in selected_versions.items():
            info = package_info.get(name, {})
            lock_data["packages"][name] = {
                "version": version,
                "language": info.get("language", "unknown"),
                "source": info.get("source", "unknown"),
                "dependencies": dependency_map.get(name, []),
                "integrity": info.get("integrity")
            }
        
        try:
            # write atomically: write to temp file first, then rename
            import tempfile
            temp_file = self.lockfile_path.with_suffix('.tmp')
            try:
                with open(temp_file, 'w') as f:
                    json.dump(lock_data, f, indent=2)
                # atomic rename
                temp_file.replace(self.lockfile_path)
                return True
            except Exception as e:
                # cleanup temp file on error
                if temp_file.exists():
                    try:
                        temp_file.unlink()
                    except Exception:
                        pass
                import logging
                logger = logging.getLogger(__name__)
                logger.error(f"Failed to write lock file: {e}")
                return False
        except Exception as e:
            import logging
            logger = logging.getLogger(__name__)
            logger.error(f"Failed to write lock file: {e}")
            return False
    
    def get_locked_versions(self) -> Dict[str, str]:
        """get locked versions from lock file"""
        lock_data = self.read()
        if not lock_data or "packages" not in lock_data:
            return {}
        
        versions = {}
        for name, pkg_data in lock_data["packages"].items():
            if isinstance(pkg_data, dict) and "version" in pkg_data:
                versions[name] = pkg_data["version"]
        
        return versions
    
    def get_locked_packages(self) -> List[str]:
        """get list of locked package names"""
        lock_data = self.read()
        if not lock_data or "packages" not in lock_data:
            return []
        
        return list(lock_data["packages"].keys())

File: cpp_implementation/src/installer/lockfile.cpp
#include "lockfile.hpp"
#include "../core/logger.hpp"
#include <iomanip>

using namespace std;

#define LOG_ERROR(msg) Logger::instance().error(msg)

const string LockFile::LOCK_FILENAME = "dpm.lock";

LockFile::LockFile(const string& directory) {
    lock_path = (fs::path(directory) / LOCK_FILENAME).string();
}

bool LockFile::exists() const {
    return fs::exists(lock_path);
}

bool LockFile::load() {
    if (!exists()) {
        return false;
    }
    
    try {
        ifstream file(lock_path);
        if (!file.is_open()) {
            LOG_ERROR("Failed to open lock file: " + lock_path);
            return false;
        }
        
        json j;
        file >> j;
        return fromJson(j);
    } catch (const exception& e) {
        LOG_ERROR("Failed to parse lock file: " + string(e.what()));
        return false;
    }
}

bool LockFile::save() const {
    try {
        // Create directory if needed
        fs::path dir = fs::path(lock_path).parent_path();
        if (!dir.empty() && !fs::exists(dir)) {
            fs::create_directories(dir);
        }
        
        ofstream file(lock_path);
        if (!file.is_open()) {
            LOG_ERROR("Failed to create lock file: " + lock_path);
            return false;
        }
        
        file << setw(2) << toJson() << endl;
        return true;
    } catch (const exception& e) {
        LOG_ERROR("Failed to write lock file: " + string(e.what()));
        return false;
    }
}

optional<LockedPackage> LockFile::getPackage(const string& name) const {
    auto it = packages.find(name);
    if (it != packages.end()) {
        return it->second;
    }
    return nullopt;
}

map<string, string> LockFile::getLockedVersions() const {
    map<string, string> versions;
    for (const auto& [name, pkg] : packages) {
        versions[name] = pkg.version;
    }
    return versions;
}

void LockFile::setFromResolution(const map<string, string>& selected_versions,
                                 const map<string, vector<string>>& dependencies,
                                 const map<string, pair<string, string>>& package_info) {
    packages.clear();
    
    for (const auto& [name, version] : selected_versions) {
        LockedPackage pkg;
        pkg.name = name;
        pkg.version = version;
        
        // Get language and source
        auto info_it = package_info.find(name);
        if (info_it != package_info.end()) {
            pkg.language = info_it->second.first;
            pkg.source = info_it->second.second;
        }
        
        // Get dependencies
        auto deps_it = dependencies.find(name);
        if (deps_it != dependencies.end()) {
            pkg.dependencies = deps_it->second;
        }
        
        packages[name] = pkg;
    }
}

bool LockFile::matches(const map<string, string>& selected_versions) const {
    if (packages.size() != selected_versions.size()) {
        return false;
    }
    
    for (const auto& [name, version] : selected_versions) {
        auto it = packages.find(name);
        if (it == packages.end() || it->second.version != version) {
            return false;
        }
    }
    return true;
}

json LockFile::toJson() const {
    json j;
    j["version"] = 1;  // Lock file format version
    j["packages"] = json::object();
    
    for (const auto& [name, pkg] : packages) {
        json pkg_json;
        pkg_json["version"] = pkg.version;
        pkg_json["language"] = pkg.language;
        pkg_json["source"] = pkg.source;
        pkg_json["dependencies"] = pkg.dependencies;
        if (!pkg.integrity.empty()) {
            pkg_json["integrity"] = pkg.integrity;
        }
        j["packages"][name] = pkg_json;
    }
    
    return j;
}

bool LockFile::fromJson(const json& j) {
    try {
        packages.clear();
        
        if (!j.contains("packages") || !j["packages"].is_object()) {
            return false;
        }
        
        for (auto& [name, pkg_json] : j["packages"].items()) {
            LockedPackage pkg;
            pkg.name = name;
            pkg.version = pkg_json.value("version", "");
            pkg.language = pkg_json.value("language", "");
            pkg.source = pkg_json.value("source", "");
            pkg.integrity = pkg_json.value("integrity", "");
            
            if (pkg_json.contains("dependencies") && pkg_json["dependencies"].is_array()) {
                for (const auto& dep : pkg_json["dependencies"]) {
                    pkg.dependencies.push_back(dep.get<string>());
                }
            }
            
            packages[name] = pkg;
        }
        
        return true;
    } catch (const exception& e) {
        LOG_ERROR("Failed to parse lock file JSON: " + string(e.what()));
        return false;
    }
}

